src/transformStateWithClones.js

```diff
@@ -7,7 +7,32 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
-}
+  const states = [];
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    let newState = { ...currentState };
+
+    switch (action.type) {
+      case 'clear':
+        newState = {};
+        break;
+      case 'addProperties':
+        Object.assign(newState, action.extraData);
+        break;
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          delete newState[key];
+        }
+        break;
+      default:
+        break;
+    }
 
+    states.push(newState);
+    currentState = newState;
+  }
+
+  return states;
```

Comment:
The variable name `newState` does not fully adhere to the naming conventions suggested. A better name would be `stateClone` or `currentStateClone` to indicate that it's a copy of the current state.